src/udp.rs

```diff
@@ -349,142 +289,133 @@ impl Inner {
-            .sum();
-        if dropped_msgs > 0 {
-            trace!(
-                "Insufficient bandwidth. Dropping {} messages with priority >= {}.",
-                dropped_msgs,
-                expired_keys[0]
-            );
+        let _ = self.out_queue.drop_expired();
+        if let Some((msg, priority)) = msg {
+            self.out_queue.push(serialise(&msg)?, priority);
         }
+        self.flush_write_until_would_block()
+    }
 
-        if let Some((msg, priority)) = msg {
-            let entry = self
-                .write_queue
-                .entry(priority)
-                .or_insert_with(|| VecDeque::with_capacity(10));
-            entry.push_back((Instant::now(), serialise(&msg)?));
+    fn write_to<T: Serialize>(&mut self, msg: Option<(T, SocketAddr, Priority)>) -> ::Res<bool> {
+        let _ = self.out_queue2.drop_expired();
+        if let Some((msg, peer, priority)) = msg {
+            self.out_queue2.push((serialise(&msg)?, peer), priority);
         }
+        self.flush_write_to_until_would_block()
+    }
 
+    /// Returns `Ok(false)`, if write to the underlying stream would block.
+    fn flush_write_until_would_block(&mut self) -> ::Res<bool> {
         loop {
-            let data = if let Some(data) = self.current_write.take() {
+            let data = if let Some(data) = self
+                .current_write
+                .take()
+                .or_else(|| self.out_queue.next_msg())
+            {
                 data
             } else {
-                let (key, (_time_stamp, data), empty) = match self.write_queue.iter_mut().next() {
-                    Some((key, queue)) => (*key, unwrap!(queue.pop_front()), queue.is_empty()),
-                    None => return Ok(true),
-                };
-                if empty {
-                    let _ = self.write_queue.remove(&key);
-                }
-                data
+                return Ok(true);
             };
 
-            match self.sock.send(&data) {
-                Ok(bytes_txd) => {
-                    if bytes_txd < data.len() {
-                        warn!(
-                            "Partial datagram sent. Will likely be interpreted as corrupted.
-                               Queued to be sent again."
-                        );
-                        self.current_write = Some(data);
-                    }
-                }
-                Err(error) => {
-                    if error.kind() == ErrorKind::WouldBlock
-                        || error.kind() == ErrorKind::Interrupted
-                    {
-                        self.current_write = Some(data);
-                        return Ok(false);
-                    } else {
-                        return Err(From::from(error));
-                    }
-                }
-            }
+            let res = self.sock.send(&data);
+            handle_send_res(res, &mut self.current_write, data.len(), data)?;
```

Comment:
Why is it not breaking out of the loop for `Ok(false)` ? At that point you will have triggered a wouldblock so you shouldn't really be attempting to do anything with the socket until the next edge-trigger for write readiness.

Do check though if that would require any other change.

Reply:
Good catch, thx :)